app/scripts/lib/account-tracker.js

```diff
@@ -158,90 +280,148 @@ export default class AccountTracker {
+    Object.keys(accountsByChainId).forEach((chainId) => {
+      addresses.forEach((address) => {
+        delete accountsByChainId[chainId][address];
+      });
+    });
     // save accounts state
-    this.store.updateState({ accounts });
+    this.store.updateState({ accounts, accountsByChainId });
   }
 
   /**
    * Removes all addresses and associated balances
    */
 
   clearAccounts() {
-    this.store.updateState({ accounts: {} });
+    const { accountsByChainId } = this.store.getState();
+    Object.keys(accountsByChainId).forEach((chainId) => {
+      accountsByChainId[chainId] = {};
+    });
+    this.store.updateState({ accounts: {}, accountsByChainId });
   }
 
   /**
-   * Given a block, updates this AccountTracker's currentBlockGasLimit, and then updates each local account's balance
-   * via EthQuery
+   * Given a block, updates this AccountTracker's currentBlockGasLimit and currentBlockGasLimitByChainId and then updates
+   * each local account's balance via EthQuery
    *
    * @private
    * @param {number} blockNumber - the block number to update to.
    * @fires 'block' The updated state, if all account updates are successful
    */
   async _updateForBlock(blockNumber) {
-    this._currentBlockNumber = blockNumber;
+    await this._updateForBlockByNetworkClientId(null, blockNumber);
+  }
+
+  /**
+   * Given a block, updates this AccountTracker's currentBlockGasLimitByChainId, and then updates each local account's balance
+   * via EthQuery
+   *
+   * @private
+   * @param {string} networkClientId - optional network client ID to use instead of the globally selected network.
+   * @param {number} blockNumber - the block number to update to.
+   * @fires 'block' The updated state, if all account updates are successful
+   */
+  async _updateForBlockByNetworkClientId(networkClientId, blockNumber) {
+    const { chainId, provider } =
+      this.#getCorrectNetworkClient(networkClientId);
+    this._currentBlockNumberByChainId[chainId] = blockNumber;
 
     // block gasLimit polling shouldn't be in account-tracker shouldn't be here...
-    const currentBlock = await this._query.getBlockByNumber(blockNumber, false);
+    const currentBlock = await pify(new EthQuery(provider)).getBlockByNumber(
+      blockNumber,
+      false,
+    );
     if (!currentBlock) {
       return;
     }
     const currentBlockGasLimit = currentBlock.gasLimit;
-    this.store.updateState({ currentBlockGasLimit });
+    const { currentBlockGasLimitByChainId } = this.store.getState();
+    currentBlockGasLimitByChainId[chainId] = currentBlockGasLimit;
+    this.store.updateState({ currentBlockGasLimitByChainId });
+    if (chainId === this.getCurrentChainId()) {
+      this.store.updateState({
+        currentBlockGasLimit,
+      });
+    }
 
     try {
-      await this._updateAccounts();
+      await this._updateAccounts(networkClientId);
     } catch (err) {
       log.error(err);
     }
   }
 
+  async _updateAccountsAllActiveNetworks() {
```

Comment:
[nit] I guess to be consistent this should have jsdocs? Its private so generally we don't require them but a bunch of other private methods on this controller have jsdocs?

Reply:
done https://github.com/MetaMask/metamask-extension/pull/21868/commits/96ddb2e3344a195dfdb9d1ac7f0a3aec395b349d